syntax = "proto3";

option optimize_for = LITE_RUNTIME;

package neptune.packets;

enum eDS {    
    _INVALID_DS_        = 0;

    ED25519             = 1;
    ED448               = 2;
    
    DILITHIUM_1024x768  = 3;
    DILITHIUM_1280x1024 = 4;
    DILITHIUM_1536x1280 = 5;
    
    FALCON_512          = 6;
    FALCON_1024         = 7;
}

enum eHashing {
    _INVALID_H_       = 0;

    SHA2_256          = 1;

    SHA3_256          = 4;

    BLAKE3_256        = 7;
}

message DSPublicKey {    
    eDS type = 1;
    bytes key = 2;
}

message DSResult {
    eDS type = 1;
    bytes result = 2;
}

enum eKEM {
    _INVALID_KEM_  = 0;

    X25519         = 1;
    X448           = 2;
        
    KYBER_512      = 3;
    KYBER_768      = 4;
    KYBER_1024     = 5;
    
    LightSABER     = 6;
    SABER          = 7;
    FireSABER      = 8;

    NTRU           = 9;
}

message KEMPublicKey {    
    eKEM type = 1;    
    bytes key = 2;
}

message Handshake {
    uint32 version = 1;
    
    repeated KEMPublicKey kem_keys = 2;
    repeated DSPublicKey ds_keys = 3;
    
    repeated DSResult ds_results = 4;
    
    bytes salt = 5;
}
